### ISSUE TITLE
Cannot establish client and server connection for Web Serving benchmark inside QEMU-KVM

### ISSUE THREAD

**Issue by Arnab035** [parsa-epfl/cloudsuite#129]
(If it matters, I am using Linux Kernel 4.4.0-62-generic on x86_64 architecture using a recent version of QEMU (2.11.92)running in KVM mode)

Hello all,
I have been facing an issue when I run the Web Serving benchmark inside of a Virtual Machine in QEMU-KVM. I can start all of the docker images associated with the Web Service benchmark normally.
I use the below commands to start all of the services -

```
docker run -dt --net=host --name=mysql_server cloudsuite/web-serving:db_server 10.0.2.15

docker run -dt --net=host --name=memcache_server cloudsuite/web-serving:memcached_server

docker run -dt --net=host --name=web_server cloudsuite/web-serving:web_server /etc/bootstrap.sh 127.0.0.1 127.0.0.1 80 (I have tried with lesser children as well)

```
Finally I run the faban_client with the below command -

`docker run --net=host --name=faban_client cloudsuite/web-serving:faban_client  10.0.2.15 7
`
(I tried using 127.0.0.1 for the Web-server IP address while launching the client as well)

Here 10.0.2.15 is the first IP address that is assigned to QEMU via DHCP. I have tried various combinations but I am still not able to fully run the Web-server benchmark because the client cannot establish connection to the server.

However, using the same set of steps I can establish connection to the server when I do not use a VM (i.e. in a bare metal environment).

I decided to look around the error messages-

Inside of the server's log file, I found error messages like these-

/var/log/nginx/examples_error.log

`recv() failed (104 : Connection reset by peer) while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "HEAD / HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "10.0.2.15:8080"`

Moreover, when I try to do a `wget` to the server from inside of the VM, I get the below message-

`wget http://10.0.2.15:8080/ `

Connecting to 10.0.2.15:8080... connected
HTTP Request sent, awaiting response... 502 Bad Gateway

So I get a Bad Gateway but the connection gets established.

I cannot seem to figure out what could be the source of this problem. Is there any way to identify and solve the cause of this problem ?

Thank You for your time.
Best Regards,
Arnab

**Comment by Hnefi** [comment]
Hi Arnab:
This is an advanced use case that will require some more thinking on our end, since there are 2 levels of virtualization involved now (QEMU/KVM and docker running inside the VMs). Just a few additional questions for you:

- Are you running all of these containers on the same VM?
- What is your DHCP setup? I am assuming that you are running the DHCP daemon on the host server and it is assigning your VMs IPs?
- Which QEMU networking frontend are you using?
- Using host network inside the VM means that all of the traffic from the containers should pass through the emulated I/O interface exposed by QEMU. However, if the docker stack is not working harmoniously with that emulated interface, it can cause problems.

A 502 error is likely meaning that the networking setup is the issue, as it signals that the entire HTTP request was not successfully completed.

**Comment by Arnab035** [comment]
Hi Mark,
Thank you for your reply.

- I use the same VM to run all of the Docker images (`webserver`, `client`,  `memcache`, `mysql_server`).
- No the guest OS inside the VM will set up a virtual DHCP server that will assign IPs to the VM. The VMs will be assigned IPs starting from 10.0.2.15.
- I use the user-mode networking feature inside of QEMU.  This will allow the guest VMs access to the internet but the ICMP protocol will not work inside of the VM (normal TCP/UDP works fine). This could be a helpful link to know - ( https://en.wikibooks.org/wiki/QEMU/Networking ) if you are unaware.
- The fact that we use `--net=host` when we run the docker images imposes the host network for the Docker containers. And yes, it could be that a different networking setup will be needed.

Note:
We tried investigating into more details about the issue and when we tried looking into the nginx server logs - we got repeated messages like this 
`child #n terminated because of SIGBUS errors: core dumped` (where n > 1) 

Regards,
Arnab
